server/auth/auth.go

```diff
@@ -657,44 +657,73 @@ func (auth *AuthManager) UserSettlingLimit(user account.AccountID, mkt *dex.Mark
+	if matchOutcomes != nil {
+		matchCounts := matchOutcomes.binViolations()
+		for v, count := range matchCounts {
 			score += v.Score() * int32(count)
 		}
+		successCount = int32(matchCounts[ViolationSwapSuccess])
 	}
-	if outcomes, found := auth.preimgOutcomes[user]; found {
-		score += ViolationPreimageMiss.Score() * outcomes.misses()
+	if preimgOutcomes != nil {
+		piMissCount = preimgOutcomes.misses()
+		score += ViolationPreimageMiss.Score() * piMissCount
 	}
 	return
 }
 
-func (auth *AuthManager) registerMatchOutcome(user account.AccountID, misstep NoActionStep, mmid db.MarketMatchID, value uint64, refTime time.Time) int32 {
+// userScore computes an authenticated user's score from their recent match
+// outcomes and preimage history. They must have entries in the outcome maps.
+// Use loadUserScore to compute score from history in DB. This must be called
+// with the violationMtx locked.
+func (auth *AuthManager) userScore(user account.AccountID) (score int32) {
+	score, _, _ = integrateOutcomes(auth.matchOutcomes[user], auth.preimgOutcomes[user])
+	return score
+}
+
+func (auth *AuthManager) registerMatchOutcome(user account.AccountID, misstep NoActionStep, mmid db.MarketMatchID, value uint64, refTime time.Time) (score int32) {
 	violation := misstep.Violation()
 
 	auth.violationMtx.Lock()
 	defer auth.violationMtx.Unlock()
-	outcomes, found := auth.matchOutcomes[user]
-	if !found {
-		outcomes = newLatestMatchOutcomes(scoringMatchLimit)
-		auth.matchOutcomes[user] = outcomes
-	}
-	outcomes.add(&matchOutcome{
-		time:    encode.UnixMilli(refTime),
-		mid:     mmid.MatchID,
-		outcome: violation,
-		value:   value,
-		base:    mmid.Base,
-		quote:   mmid.Quote,
-	})
+	matchOutcomes, found := auth.matchOutcomes[user]
+	if found {
+		matchOutcomes.add(&matchOutcome{
+			time:    encode.UnixMilli(refTime),
+			mid:     mmid.MatchID,
+			outcome: violation,
+			value:   value,
+			base:    mmid.Base,
+			quote:   mmid.Quote,
+		})
+		score = auth.userScore(user)
+		log.Debugf("Registering outcome %q (badness %d) for user %v, new score = %d",
+			violation.String(), violation.Score(), user, score)
+		return
+	}
+
+	// The user is currently not connected and authenticated. When the user logs
+	// back in, their history will be reloaded (loadUserScore) and their account
+	// will be suspended/restored as required, but compute their score now from
+	// DB so their orders may be unbooked if need.
+	matchOutcomes, piOutcomes, err := auth.loadUserOutcomes(user)
+	if err != nil {
+		log.Errorf("Failed to load swap and preimage outcomes for user %v: %v", user, err)
+		return 0
+	}
+
+	// Make outcome entries for the user to optimize subsequent outcomes calls
+	// while they are disconnected? This could lead to adding duplicate outcomes
+	// with a concurrent connect/login or subsequent outcomes while offline.
+	//
+	// auth.matchOutcomes[user] = matchOutcomes
+	// auth.preimgOutcomes[user] = piOutcomes
```

Comment:
This would be safe with the new duplicate checks in the outcome structs' `add` methods, but it's a little messy and these map entries are assumed not to be present for a disconnected user.

Even for the largest matches table so far, the `loadUserOutcomes` query is ~200ms on my (admittedly fast) machine. Even the occasional string of a dozen back-to-back outcomes for offline users (match failures or preimage misses) should be tolerable. We can experiment with this optimization or even separate outcome maps if it becomes slow to hit the DB for each outcome for an offline user.